### DISCUSSION TITLE
mupdf: cannot access files in whitelisted ~/Desktop: Permission denied

### DISCUSSION THREAD

**Discussion by wonbug** [netblue30/firejail#4975] [Q&A] [answered]
I'm trying to open a PDF file with `mupdf` in `~/Desktop/`

I put all these directives into `~/.config/firejail/mupdf.local`:

```
whitelist ${HOME}/Desktop/
whitelist ${HOME}/Desktop
whitelist ~/Desktop/
whitelist ~/Desktop
```

Yet when I try to open the file, I get:

```
$ /usr/bin/firejail /usr/bin/mupdf ~/Desktop/file.pdf
...
error: cannot open /home/user/Desktop/file.pdf: Permission denied
mupdf: error: cannot open document
```

Am I missing something obvious?

**Comment by ghost** [comment]
Firejail refers to that path as ${DESKTOP}, for example in [disable-xdg.inc](https://github.com/netblue30/firejail/blob/master/etc/inc/disable-xdg.inc#L11). You only need `whitelist ${DESKTOP}` in your mupdf.local in this case.

**Reply by rusty-snake** [reply]
> `whitelist ${DESKTOP}`

Is the same as in OP (or at least should be the same). The problem is something else.

**Reply by wonbug** [reply]
No dice with the ${DESKTOP} directive either.

```
$ cat ~/.config/firejail/mupdf.local
whitelist ${DESKTOP}
```

Same "No such file or directory" file error as before.

**Comment by wonbug** [comment]
Stating the obvious: file permissions are fine and using `/usr/bin/mupdf` directly instead of `firejail /usr/bin/mupdf` opens the file fine.

**Comment by ghost** [comment]
Installed [mupdf](https://archlinux.org/packages/community/x86_64/mupdf/) here just now. Put a PDF at ~/Desktop/file.pdf and without using a mupdf.local the file opens just fine when running

$ /usr/bin/firejail /usr/bin/mupdf ~/Desktop/file.pdf

Are you trying to implement a whitelisting profile for mupdf? Absolutely sure you don't have a disable-xdg.local that might interfer? It would help us help you if you could post output from the command(s) you're trying. Using `--ignore=quiet` for example will show the list of profile/local files being processed.

One thing I noticed is that /etc/firejail/mupdf.profile has`#include globals.local`, which IMO is a bug, but that shouldn't be causing this issue.

**Reply by wonbug** [reply]
firejail does load `/etc/firejail/disable-xdg.inc` but that file does not disallow access to Desktop, and shouldn't a local config overwrite it even if it did?

```
Reading profile /etc/firejail/mupdf.profile
Reading profile /home/user/.config/firejail/mupdf.local
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /home/user/.config/firejail/disable-programs.local
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 11681, child pid 11682
Private /etc installed in 20.89 ms
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Blacklist violations are logged to syslog
Child process initialized in 123.13 ms
error: cannot open Desktop/file.pdf: Permission denied
mupdf: error: cannot open document

Parent is shutting down, bye...
```

**Comment by crocket** [comment]
Have you enabled apparmor or SELinux? AppArmor and SELinux can deny permissions.

**Reply by rusty-snake** [reply]
> using `/usr/bin/mupdf` directly instead of `firejail /usr/bin/mupdf` opens the file fine.

**Reply by crocket** [reply]
firejail-default apparmor profile could have caused permission denial.

**Reply by wonbug** [reply]
`selinuxenabled` returns 1, so I believe it is off.

**Reply by crocket** [reply]
Did you fix the issue already?

**Reply by wonbug** [reply]
No, still happens.

**Comment by ghost** [comment]
Just to rule out things we don't know about yet, please post on which OS you're seeing this + firejail version. Also, run the below command and show us the resulting debug file (it will be placed in ~/Downloads/fj-mupdf-debug):

```
$ firejail --debug /usr/bin/mupdf ~/Desktop/file.pdf | tee -a ~/Downloads/fj-mupdf-debug > /dev/null
```

**Comment by wonbug** [comment]
```
$ which firejail
/usr/local/bin/firejail

$ firejail --version
firejail version 0.9.67

$ cat .config/firejail/mupdf.local
whitelist ${DESKTOP}

$ firejail --debug /usr/bin/mupdf ~/Desktop/file.pdf | tee -a ~/Downloads/fj-mupdf-debug > /dev/null
Reading profile /usr/local/etc/firejail/mupdf.profile
Reading profile /home/user/.config/firejail/mupdf.local
Reading profile /usr/local/etc/firejail/disable-common.inc
Reading profile /usr/local/etc/firejail/disable-devel.inc
Reading profile /usr/local/etc/firejail/disable-exec.inc
Reading profile /usr/local/etc/firejail/disable-interpreters.inc
Reading profile /usr/local/etc/firejail/disable-programs.inc
Reading profile /home/user/.config/firejail/disable-programs.local
Reading profile /usr/local/etc/firejail/disable-xdg.inc
Reading profile /usr/local/etc/firejail/whitelist-var-common.inc
[profile] combined protocol list: "unix"
DISPLAY=:0 parsed as 0
Parent pid 9517, child pid 9519
Private /etc installed in 23.02 ms
Private /usr/etc installed in 0.02 ms
nogroups command not ignored
Warning: cleaning all supplementary groups
nogroups command not ignored
Warning: cleaning all supplementary groups
DISPLAY=:0 parsed as 0
Blacklist violations are logged to syslog
nogroups command not ignored
nogroups command not ignored
Warning: cleaning all supplementary groups
Child process initialized in 137.03 ms
error: cannot open /home/user/Desktop/file.pdf: Permission denied
mupdf: error: cannot open document

Parent is shutting down, bye...
```

The debug stdout doesn't reveal anything obvious. I can post the whole thing but it's several thousand lines. The path looks like it resolves correctly:

```
Debug 558: whitelist ${DESKTOP}
Directory ${DESKTOP} resolved as Desktop
Debug 579: expanded: /home/user/Desktop
Debug 590: new_name: /home/user/Desktop
```

I'm extra confused now because firejail mupdf no longer opens files in ~/Downloads or any other directory in HOME (same "no such file" error), but if I remove the `whitelist ${DESKTOP}` directive from mupdf.local, it resumes working. Huh?

**Reply by wonbug** [reply]
I'm on Debian 11.2 (Linux 5.10) by the way.

**Reply by ghost** [reply]
I'm not familiar with Debian's Firejail support, but the most current release is [0.9.68](https://github.com/netblue30/firejail/releases/tag/0.9.68). As usual it is advised to run the latest release, which has bug fixes, newly introduced profiles, etcetera. If you can, try the [backports](https://packages.debian.org/bullseye-backports/firejail) package as mentioned [here](https://github.com/netblue30/firejail#debian).

> I'm extra confused now because firejail mupdf no longer opens files in ~/Downloads or any other directory in HOME (same "no such file" error), but if I remove the whitelist ${DESKTOP} directive from mupdf.local, it resumes working. Huh?

Sounds confusing. But, in fact, this is what I mentioned in https://github.com/netblue30/firejail/discussions/4975#discussioncomment-2247633: nothing extra is needed to open a file in ~/Desktop with mupdf because ${DESKTOP} is never blacklisted anywhere. You _can_ turn the default mupdf.profile into what is called a `whitelisting` profile to harden it further. That turns the 'logic' upside-down and paths that aren't explicitly whitelisted are made inaccessible in the sandbox:

```console
$ cat ~/.config/firejail/mupdf.local
whitelist ${DESKTOP}
#whitelist ${DOWNLOADS}
include whitelist-common.inc
include whitelist-run-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
```

If you try to open ~/Downloads/file.pdf with the above local override, mupdf will complain that it cannot open that path.

Maybe you had a leftover in your ~/.config/firejail/mupdf.local that triggered the blocking before, I don't know.

**Comment by crocket** [comment]
The culprit is disable-xdg.inc. disable-xdg has
```
#blacklist ${DESKTOP}
```
```
--ignore="include disable-xdg.inc"
```
fixes the issue. I don't know how a commented line seems to have effect.

**Comment by ghost** [comment]
@crocket That would be quite a bug. What happens when you remove all the commented blacklist lines from disable-xdg.inc?

**Reply by crocket** [reply]
I removed comments at the bottom without seeing any change.
If `blacklist ${MUSIC}`, `blacklist ${PICTURES}`, or `blacklist ${VIDEOS}` is not commented out in disable-xdg.inc, `${DESKTOP}` is blacklisted.
`blacklist ${DOCUMENTS}` doesn't blacklist `${DESKTOP}`.

**Reply by ghost** [reply]
Thanks for the info. We'd better turn this into an issue to attract more attention. Could be a nasty bug that we need to look at. Ever since we have `disable-xdg.inc` I created a `disable-xdg.local` that has `blacklist ${DESKTOP}` because I use the ~/Desktop directory for other stuff and never want it accessible inside sandboxes. Perhaps that's why I cannot reproduce. A bit pressed for time at the moment, but definately something strange is going on here.

**Comment by ghost** [comment, marked as answer]
@wonbug Turns out this needs further investigation. Let's move the discussion to https://github.com/netblue30/firejail/issues/4994.

**Reply by smitsohu** [reply]
@wonbug There is a tentative [fix](https://github.com/netblue30/firejail/commit/a146a985b9d6142b57d78b1bf9a4ab461b7f10dc) now in Firejail master, if you find the time could yo/u confirm it resolves your issue? Thanks.

Let us know you need assistance with this.

**Reply by wonbug** [reply]
I pulled down the latest source, checked for your change is in `macros.c`, ran `make` then `sudo make install`

Opened a new terminal, `mupdf Desktop/file.pdf` still results in the same error, with or without the `whitelist ${DESKTOP}` in `~/.config/firejail/mupdf.local`

I tried to run `firecfg` but now appear unable to?

```
$ sudo firecfg
Removing all firejail symlinks:

fopen: No such file or directory
Error: cannot open /usr/local/etc/firejail/firecfg.config
```

**Reply by smitsohu** [reply]
> I tried to run firecfg but now appear unable to?

It looks like there is something wrong with your installation. Can you re-`configure` and `make` and install again?

> Opened a new terminal, mupdf Desktop/file.pdf still results in the same error, with or without the whitelist ${DESKTOP} in ~/.config/firejail/mupdf.local

Thanks. There is nothing obviously wrong in what you are doing. But I have the impression that we need a simpler reproducer. Can you confirm that these commands work for you: 
`firejail --profile=/usr/local/etc/firejail/disable-xdg.inc mupdf ~/Desktop/file.pdf` and
`firejail --noprofile '--whitelist=${DESKTOP}' mupdf ~/Desktop/file.pdf` 

If yes, could you comment/uncomment lines in mupdf.profile in order to find the line that breaks it?
